tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/WriteSourceRepoProperties.cs

```diff
@@ -0,0 +1,260 @@
+            }
+            return repoUrl.Substring(repoUrl.LastIndexOf("/") + 1);
+        }
+
+        private static string GetCommitCount(string gitDir, string hash)
+        {
+            return RunGitCommand($"rev-list --count {hash}", gitDir: gitDir);
+        }
+
+        private static string GetCommitDate(string gitDir, string hash)
+        {
+            return RunGitCommand($"log -1 --format=%cd --date=short", gitDir: gitDir);
+        }
+
+        private static string RunGitCommand(string command, string workTree = null, string gitDir = null)
+        {
+            // Windows Git requires these to be before the command
+            if (workTree != null)
+            {
+                command = $"--work-tree={workTree} {command}";
+            }
+            if (gitDir != null)
+            {
+                command = $"--git-dir={gitDir} {command}";
+            }
+            var psi = new ProcessStartInfo { FileName = "git", Arguments = command, RedirectStandardOutput = true, RedirectStandardError = true, UseShellExecute = false, CreateNoWindow = true };
+            var p = new Process { StartInfo = psi };
+            p.Start();
+            var output = p.StandardOutput.ReadToEnd();
+            var error = p.StandardError.ReadToEnd();
+            p.WaitForExit();
+            if (p.ExitCode != 0)
+            {
+                throw new InvalidOperationException($"git command {command} failed with exit code {p.ExitCode} and error {error ?? "<blank>"}");
+            }
+            return output.Trim();
+        }
+
+        private static string DeriveRepoGitDirPath(string gitDirPath, string repoUrl)
+        {
+            if (repoUrl.EndsWith(".git"))
+            {
+                repoUrl = repoUrl.Substring(0, repoUrl.Length - ".git".Length);
+            }
+
+            return Path.Combine(gitDirPath, $"{repoUrl.Substring(repoUrl.LastIndexOf("/") + 1)}.git");
+        }
+
+        private static void HandleSubmodules(string sourceDirPath, string gitDirPath, Dependency dependency)
+        {
+            var gitModulesPath = Path.Combine(sourceDirPath, ".gitmodules");
+            var gitModules = File.ReadAllText(gitModulesPath);
+            // This is a fun one.
+            // Start with header [submodule "name"]
+            // Then we have any number of parameters we don't care about, like
+            //      branch = release/3.0
+            //      url = https://example.com/project.git
+            // Then we have the one we do care about:
+            //      path = some/file/path
+            // Then we have possibly more stuff we don't care about.
+            // And finally we non-consuming anchor to either the next submodule header "[" or the end of the input so we stop capturing.
+            var gitModuleRegex = new Regex(@"\[\s*submodule\s+""(?<submoduleName>.+?)""\](\s+\w+\s*=\s*(.+?)\s)*?\s+path\s*=\s*(?<submodulePath>.+?)\s(\w+\s*=\s*(.+?)\s)*?(?=(\[|$))", RegexOptions.Singleline);
```

Comment:
Quite something... considered `git config --file {gitModules} --get submodule.{submoduleName}.path`?

Reply:
We don't actually know the submodule names at this point, but I was able to simplify this massively with `git config --file={gitModulesFilePath} --list` and trimming the regex to just `submodule\.(?<submoduleName>.*)\.path=(?<submodulePath>.*)`, thanks!